### PR TITLE
Add async counterparts to S3Client::upload and S3Client::copy

### PR DESCRIPTION
**Pull request by jeskew**
This PR adds `uploadAsync` and `copyAsync` methods to `Aws\S3\S3Client`, both of which return promises that will be fulfilled with the result of the underlying operation. `upload` and `copy` now simply call `uploadAsync` and `copyAsync` respectively and wait on the returned promise.

/cc @mtdowling

### CODE REVIEW

File: src/S3/S3Client.php [hunk 1]
```diff
@@ -467,43 +500,76 @@ public function copy(
...
-        if ($maxCopyableSize < $objectStats['ContentLength']) {
-            return (new MultipartCopy($this, $copySource, [
-                'bucket' => $bucket,
-                'key' => $key,
-                'source_metadata' => $objectStats,
-                'acl' => $acl,
-            ] + $options))
-                ->upload();
-        }
+        return $this->headObjectAsync($headParams)
```
Comment (jeremeamia) [inline src/S3/S3Client.php]: Not sure, but maybe a coroutine would make this easier to read/understand.

Comment (fntlnz) [inline src/S3/S3Client.php]: :+1:  for the coroutine idea!

File: src/S3/MultipartCopy.php [hunk 2]
```diff
@@ -57,6 +57,11 @@ public function __construct(S3Client $client, $source, array $config = [])
         ]);
     }
 
+    public function copy()
```
Comment (mtdowling) [inline src/S3/MultipartCopy.php]: Missing docblock

File: src/S3/S3Client.php [hunk 3]
```diff
@@ -467,43 +501,81 @@ public function copy(
...
+     *
+     * @see self::copy for more info about the parameters above.
+     * @return PromiseInterface     Returns a promise that will be fulfilled
+     *                              with the result of the copy.
+     */
+    public function copyAsync(
+        $fromBucket,
+        $fromKey,
+        $bucket,
+        $key,
```
Comment (mtdowling) [inline src/S3/S3Client.php]: Can you rename these two parameters, bucket and key, to something like `$toBucket`, `$toKey` or `$destBucket`, `$destKey`?

**Comment by mtdowling** [comment]
Couple pieces of feedback, but otherwise, :shipit:

**Pull request merged**